pkg/proto/stream/stream_test.go

```diff
@@ -0,0 +1,286 @@
+	assert.Equal(t, true, res.Boo)
+	assert.Equal(t, "s", res.Str)
+	assert.Equal(t, "b", string(res.Byt))
+	assert.Equal(t, XmasTree_e2, res.Enu)
+}
+
+// Test that the zero values for each field do not result in any encoded data.
+func TestSimpleEncodingZero(t *testing.T) {
+	output := bytes.NewBuffer([]byte{})
+	ps := New(output)
+
+	require.NoError(t, ps.Double(fieldDub, 0.0))
+	require.NoError(t, ps.Float(fieldFlo, 0.0))
+	require.NoError(t, ps.Int32(fieldI32, int32(0)))
+	require.NoError(t, ps.Int64(fieldI64, int64(0)))
+	require.NoError(t, ps.Uint32(fieldU32, uint32(0)))
+	require.NoError(t, ps.Uint64(fieldU64, uint64(0)))
+	require.NoError(t, ps.Sint32(fieldS32, int32(0)))
+	require.NoError(t, ps.Sint64(fieldS64, int64(0)))
+	require.NoError(t, ps.Fixed32(fieldF32, uint32(0)))
+	require.NoError(t, ps.Fixed64(fieldF64, uint64(0)))
+	require.NoError(t, ps.Sfixed32(fieldSF32, int32(0)))
+	require.NoError(t, ps.Sfixed64(fieldSF64, int64(0)))
+	require.NoError(t, ps.Bool(fieldBoo, false))
+	require.NoError(t, ps.String(fieldStr, ""))
+	require.NoError(t, ps.Bytes(fieldByt, []byte("")))
+	require.NoError(t, ps.Int32(fieldEnu, enuE0))
+
+	buf := output.Bytes()
+	// all of those are zero values, so nothing should have been written
+	require.Equal(t, 0, len(buf))
+}
+
+// Test that the zero values for packed fields do not result in any encoded data.
+func TestPackedEncodingZero(t *testing.T) {
+	output := bytes.NewBuffer([]byte{})
+	ps := New(output)
+
+	require.NoError(t, ps.DoublePacked(fieldDub, []float64{}))
+	require.NoError(t, ps.FloatPacked(fieldFlo, []float32{}))
+	require.NoError(t, ps.Int32Packed(fieldI32, []int32{}))
+	require.NoError(t, ps.Int64Packed(fieldI64, []int64{}))
+	require.NoError(t, ps.Uint32Packed(fieldU32, []uint32{}))
+	require.NoError(t, ps.Uint64Packed(fieldU64, []uint64{}))
+	require.NoError(t, ps.Sint32Packed(fieldS32, []int32{}))
+	require.NoError(t, ps.Sint64Packed(fieldS64, []int64{}))
+	require.NoError(t, ps.Fixed32Packed(fieldF32, []uint32{}))
+	require.NoError(t, ps.Fixed64Packed(fieldF64, []uint64{}))
+	require.NoError(t, ps.Sfixed32Packed(fieldSF32, []int32{}))
+	require.NoError(t, ps.Sfixed64Packed(fieldSF64, []int64{}))
+
+	buf := output.Bytes()
+	// all of those are zero values, so nothing should have been written
+	require.Equal(t, 0, len(buf))
+}
+
+// Test that *Packed functions work.
+func TestPacking(t *testing.T) {
+	output := bytes.NewBuffer([]byte{})
+	ps := New(output)
+
+	assertBytes := func(t *testing.T, exp []byte, got []byte) {
+		assert.Equal(t, fmt.Sprintf("%#v", exp), fmt.Sprintf("%#v", got))
+	}
+
+	key := func(fieldNumber int) uint8 {
+		return uint8((fieldNumber << 3) + wtLengthDelimited)
+	}
+
+	t.Run("DoublePacked", func(t *testing.T) {
+		output.Reset()
+		require.NoError(t, ps.DoublePacked(fieldDub, []float64{3.14, 1.414}))
+		assertBytes(t, []byte{key(fieldDub), 0x10, 0x1f, 0x85, 0xeb, 0x51, 0xb8, 0x1e, 0x9, 0x40, 0x39, 0xb4, 0xc8, 0x76, 0xbe, 0x9f, 0xf6, 0x3f}, output.Bytes())
```

Comment:
Yeah, I can break that down with some comments as to what each byte means.

Protobufs are specified independent of a library or version, so there's no need to reference a specific library + version.  If we needed to update this, that would due to an incompatible change in proto (old and new payloads would not be compatible -- something protobufs exists to avoid).

Reply:
Encoding is documented [here](https://developers.google.com/protocol-buffers/docs/encoding)